### PR TITLE
Add UseWindowsForms and UseWPF properties

### LINKED ISSUE #7239: Expose UseWindowsForms and UseWPF properties
I believe these properties are only meaningful on .NET Core 3.0/3.1, but it would still be useful to have them as they control whether or not WinForms and WPF functionality is available to the project.

The Application property page seems like a natural place for these. However, we only want them to show for projects that target .NET Core 3.0/3.1.

_This work was identified as part of #6935._

### PR DESCRIPTION
**Pull request by drewnoakes**
Fixes #7239. Contributes to #7279.

The `UseWindowsForms` and `UseWPF` properties are only applicable when targeting .NET Core 3.0 or later.

Until recently, we did not have a means to express a visibility condition based upon target framework.

Now that we have one, we can add these properties to the UI and have them behave correctly.

cc @RussKie

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/project-system/pull/7281)

### CODE REVIEW

**Comment by RussKie** [comment]
Any screenshots to share?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/PropertyPages/ApplicationPropertyPage.xaml
Comment (RussKie) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/PropertyPages/ApplicationPropertyPage.xaml:111]: https://review.docs.microsoft.com/help/contribute/links-how-to?branch=master#locale-codes-in-external-links
```suggestion
                HelpUrl="https://docs.microsoft.com/dotnet/core/project-sdk/msbuild-props-desktop"
```

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/PropertyPages/ApplicationPropertyPage.xaml
Comment (RussKie) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/PropertyPages/ApplicationPropertyPage.xaml:129]: ditto

**Comment by drewnoakes** [comment]
> Any screenshots to share?

I should have captured some, but I've unravelled all the threads here in order to submit PRs. This change requires bringing together about 5 branches across two repos and three VSIX packages. It would take me a while to get them together.

But basically it's just:

```
Windows Forms
[ ] Enable Windows Forms for this project.

Windows Presentation Foundation
[ ] Enable WPF for this project.
```

And these items are only present if the `OutputType` is `WinExe` and the project targets .NET Core 3.0 or later. If multi-targeting, then only one target has to satisfy the constraint.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/PropertyPages/ApplicationPropertyPage.xaml
Comment (RussKie) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/PropertyPages/ApplicationPropertyPage.xaml:118]: You probably want to force `WinExe` over `Exe` if either of these frameworks are selected: https://review.docs.microsoft.com/help/contribute/links-how-to?branch=master#locale-codes-in-external-links

Comment (drewnoakes) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/PropertyPages/ApplicationPropertyPage.xaml:118]: Is that the right link? I'm not sure what you mean by forcing here.

Comment (RussKie) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/PropertyPages/ApplicationPropertyPage.xaml:118]: Of course not 🤦‍♂️ Here's the right link: https://docs.microsoft.com/dotnet/core/compatibility/sdk/5.0/automatically-infer-winexe-output-type#change-description

image
This what prompted the comment about forcing - e.g. is a user checked `UseWindowsForms` or `UseWPF` we shouldn't allow switching to `Exe`. May be even remove it from the list?
Likewise may be show a checkbox for `DisableWinExeOutputInference`?

Comment (drewnoakes) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/PropertyPages/ApplicationPropertyPage.xaml:118]: Interesting. Offhand I'm not sure we can dynamically refresh the contents of the output type list.

What we _could_ do is have the `UseWindowsForms` and `UseWPF` properties be deleted from the project if the user selects an output type other than `WinExe`.

Comment (RussKie) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/PropertyPages/ApplicationPropertyPage.xaml:118]: Strictly speaking Exe is an accepted value for a Windows Forms app, it is the same as WinExe as of .NET 5.0.
Also one can build a DLL targeting Windows Forms, e.g. GitUI or GitCommands are both libraries targeting Windows Forms.

**Comment by drewnoakes** [comment]
> Any screenshots to share?

image

And not present for .NET Framework:

image

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/PropertyPages/ApplicationPropertyPage.xaml
Comment (drewnoakes) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/PropertyPages/ApplicationPropertyPage.xaml]: This is awesome, thanks. Fixed in this PR, and I also filed https://github.com/dotnet/project-system/pull/7283 to fix it in a few other places. Appreciated.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/PropertyPages/ApplicationPropertyPage.xaml
Comment (tmeschter) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/PropertyPages/ApplicationPropertyPage.xaml:118]: The Property Pages UI will reflect the final evaluated value of the OutputType property. So for WinForms or WPF apps if you try to set it to Exe in the UI the project file will get updated--and then the UI will reset itself because the value is overridden in the targets.

This means the `DependsOn` is backwards--the OutputType effectively depends on `UseWindowsForms` and `UseWPF` even though for the purposes of showing these controls we would like it to be the other way around.

I agree with @drewnoakes; let's get this in and see what the feedback is. It's better to have these checkboxes in some form now rather than wait until we've figured out the perfect implementation.

**Review by tmeschter — APPROVED** [review]

**Comment by drewnoakes** [comment]
I'll merge this once the CPS bits are reviewed and merged.

> This means the `DependsOn` is backwards

I actually think we over-specify `DependsOn` in several places. It could be revisited.

**Review by MiYanni — APPROVED** [review]

**Pull request merged**
